### PR TITLE
Refactor cosine similarity and add cosine distance

### PR DESCRIPTION
**Pull request by Shashi456**
Refer to [this](https://github.com/tensorflow/swift-apis/pull/233#discussion_r293516299)

cc: @rxwei. 
Actually if we are removing the argument labels, i think we should do away with predicted and expected and have separate variable names.

### CODE REVIEW

**Comment by Shashi456** [comment]
@rxwei Ah, I'll fix this immediately. Thanks for taking time and explaining this to me.

File: Sources/TensorFlow/Loss.swift [hunk 1]
```diff
@@ -109,7 +109,7 @@ public func hingeLoss<Scalar: TensorFlowFloatingPoint>(
 ///   - expected: Expected values, i.e. targets, that correspond to the correct output.
 @differentiable(wrt: (predicted, expected))
 public func cosineSimilarity<Scalar: TensorFlowFloatingPoint>(
-    predicted: Tensor<Scalar>, expected: Tensor<Scalar>
+    _ predicted: Tensor<Scalar>, _ expected: Tensor<Scalar>
```
Comment (rxwei) [inline Sources/TensorFlow/Loss.swift]: Thanks for the change. With this, the call sites no longer need to write `predicted:` or `expected:`, which is good. However, Since this is an API for more general use than loss functions, we should not distinguish between `predicted` and `expected` at the implementation site or in API documentation comments. As such, could you please rename argument names `predicted` and `expected` to `x` and `y` respectively?

Comment (Shashi456) [reply Sources/TensorFlow/Loss.swift]: @rxwei How about I put this as a Math op and have a higher level function within losses calling the base function? would that count as redundancy of code? Because It's just sort of weird that we are putting a general API usage function in losses.swift

Comment (rxwei) [reply Sources/TensorFlow/Loss.swift]: Putting it in Math.swift is good. A wrapper in Loss.swift is unnecessary.

**Comment by Shashi456** [comment]
@rxwei well the build passes locally. I cant build tests because i keep getting error since #234. Can you trigger a build?

**Comment by Shashi456** [comment]
@rxwei Why do some kokoro labels get blocked?

File: Sources/TensorFlow/Operators/Math.swift [hunk 2]
```diff
@@ -644,6 +644,15 @@ public func rsqrt<T: TensorFlowFloatingPoint>(_ x: Tensor<T>) -> Tensor<T> {
     Raw.rsqrt(x)
 }
 
+/// Returns the cosine similarity between x and y.
```
Comment (rxwei) [inline Sources/TensorFlow/Operators/Math.swift]: Please add backquotes around all parameters (or generally any program identifiers) in API documentation comments.
```suggestion
/// Returns the cosine similarity between `x` and `y`.
```

**Comment by eaplatanios** [comment]
This is not computing cosine similarity but rather the negative cosine similarity. I would suggest you remove the minus sign and also introduce a dual function called `cosineDistance` which is computed as `1 - cosineSimilarity`.

File: Sources/TensorFlow/Operators/Math.swift [hunk 3]
```diff
@@ -644,6 +644,23 @@ public func rsqrt<T: TensorFlowFloatingPoint>(_ x: Tensor<T>) -> Tensor<T> {
     Raw.rsqrt(x)
 }
 
+/// Returns the cosine similarity between `x` and `y`.
+@differentiable(wrt: (x, y))
```
Comment (rxwei) [inline Sources/TensorFlow/Operators/Math.swift]:
```suggestion
@differentiable
```

File: Sources/TensorFlow/Operators/Math.swift [hunk 4]
```diff
@@ -644,6 +644,23 @@ public func rsqrt<T: TensorFlowFloatingPoint>(_ x: Tensor<T>) -> Tensor<T> {
...
+public func cosineSimilarity<Scalar: TensorFlowFloatingPoint>(
+    _ x: Tensor<Scalar>, _ y: Tensor<Scalar>
+) -> Tensor<Scalar> {
+    return (x * y).sum() /
+        (sqrt(x.squared().sum()) * sqrt(y.squared().sum()))
+}
+
+/// Returns the cosine distance between `x` and `y`.
+@differentiable(wrt: (x, y))
```
Comment (rxwei) [inline Sources/TensorFlow/Operators/Math.swift]:
```suggestion
@differentiable
```

File: Sources/TensorFlow/Operators/Math.swift [hunk 5]
```diff
@@ -644,6 +644,23 @@ public func rsqrt<T: TensorFlowFloatingPoint>(_ x: Tensor<T>) -> Tensor<T> {
     Raw.rsqrt(x)
 }
 
+/// Returns the cosine similarity between `x` and `y`.
+@differentiable(wrt: (x, y))
+public func cosineSimilarity<Scalar: TensorFlowFloatingPoint>(
+    _ x: Tensor<Scalar>, _ y: Tensor<Scalar>
+) -> Tensor<Scalar> {
+    return (x * y).sum() /
```
Comment (rxwei) [inline Sources/TensorFlow/Operators/Math.swift]: I think if you remove the `return` keyword the whole body can fit in one line.

```suggestion
    (x * y).sum() / (sqrt(x.squared().sum()) * sqrt(y.squared().sum()))
```

File: Sources/TensorFlow/Operators/Math.swift [hunk 6]
```diff
@@ -644,6 +644,23 @@ public func rsqrt<T: TensorFlowFloatingPoint>(_ x: Tensor<T>) -> Tensor<T> {
...
+        (sqrt(x.squared().sum()) * sqrt(y.squared().sum()))
+}
+
+/// Returns the cosine distance between `x` and `y`.
+@differentiable(wrt: (x, y))
+public func cosineDistance<Scalar: TensorFlowFloatingPoint>(
+    _ x: Tensor<Scalar>, _ y: Tensor<Scalar>
+) -> Tensor<Scalar> {
+    return 1 - cosineSimilarity(x, y)
```
Comment (rxwei) [inline Sources/TensorFlow/Operators/Math.swift]:
```suggestion
    1 - cosineSimilarity(x, y)
```

File: Sources/TensorFlow/Operators/Math.swift [hunk 7]
```diff
@@ -644,6 +644,23 @@ public func rsqrt<T: TensorFlowFloatingPoint>(_ x: Tensor<T>) -> Tensor<T> {
...
+/// Returns the cosine similarity between `x` and `y`.
+@differentiable(wrt: (x, y))
+public func cosineSimilarity<Scalar: TensorFlowFloatingPoint>(
+    _ x: Tensor<Scalar>, _ y: Tensor<Scalar>
+) -> Tensor<Scalar> {
+    return (x * y).sum() /
+        (sqrt(x.squared().sum()) * sqrt(y.squared().sum()))
+}
+
+/// Returns the cosine distance between `x` and `y`.
```
Comment (rxwei) [inline Sources/TensorFlow/Operators/Math.swift]:
```suggestion
/// Returns the cosine distance between `x` and `y`. Cosine distance is defined as
/// `1 - cosineSimilarity(x, y)`.
```

File: Sources/TensorFlow/Operators/Math.swift (hunk 3 above)
Comment (rxwei) [reply Sources/TensorFlow/Operators/Math.swift]: Hmm this doesn't seem to have been resolved yet.

**Comment by Shashi456** [comment]
Also @rxwei any idea on passing #31, one of the last losses to be added.

**Comment by Shashi456** [comment]
@rxwei could you retrigger a build?

File: Tests/TensorFlowTests/OperatorTests/MathTests.swift [hunk 8]
```diff
@@ -65,6 +65,14 @@ final class MathOperatorTests: XCTestCase {
...
     }
 
+    func testCosineSimilarity() {
+        let x = Tensor<Float>([1, 2, 3, 4, 5, 6, 7, 8])
+        let y = Tensor<Float>([0.5, 1, 1.5, 2.0, 2.5, 3.0, 3.5, 4.0])
+        let z = cosineSimilarity(x, y)
+        let output: Float = 1.0
+        XCTAssertEqual(z, Tensor(output))
+    }
+
```
Comment (saeta) [inline Tests/TensorFlowTests/OperatorTests/MathTests.swift]: Could you please also add a test for cosine distance?

Comment (Shashi456) [reply Tests/TensorFlowTests/OperatorTests/MathTests.swift]: @saeta, actually cosine distance just returns 1 - cosineSimilarity, so in effect we would be testing the same function. To avoid redundancy I didn't add a test

Comment (rxwei) [reply Tests/TensorFlowTests/OperatorTests/MathTests.swift]: Even though the function is implemented as a trivial wrapper, it's good practice to assume the implementation can change in the future for whatever reasons and have tests be implementation-agnostic.

**Comment by rxwei** [comment]
Merging this to keep things rolling. It is fine to add a test for `cosineDistance(_:_:)` later.
